feat(rectifier-builder): categorized + prioritized prompt sections

## Problem

The autofix rectification prompt bundles failures from all 6 review check categories (typecheck, lint, build, test, semantic, adversarial) into a single flat list with identical formatting and no priority guidance. Each category requires a different repair pattern — local edit for typecheck, architectural rethink for adversarial — but the prompt structure makes them visually indistinguishable.

Observed in the dogfood prompt audit at `nax-dogfood/fixtures/tdd-calc/.nax/prompt-audit/tdd-calc/1777467865587-…-implementer-rectification-t01.txt`, and in the implementation at:

- [`src/prompts/builders/rectifier-builder.ts:53-82`](https://github.com/nathapp-io/nax/blob/main/src/prompts/builders/rectifier-builder.ts#L53-L82) — `firstAttemptDelta()`
- [`src/prompts/builders/rectifier-builder.ts:91-130`](https://github.com/nathapp-io/nax/blob/main/src/prompts/builders/rectifier-builder.ts#L91-L130) — `continuation()`

Both methods iterate `failedChecks` and emit `### ${check.check} (exit ${exitCode})` blocks regardless of category, ending with `Fix ALL issues listed`.

## Failure modes this causes

1. **Cherry-picking.** Agent fixes the easy mechanical failures (typecheck/lint), hand-waves the harder LLM findings ("All checks pass" after touching only the typecheck error).
2. **Conflicting fixes.** Mechanical fix for a type error and adversarial fix for "function too long" interact — the agent does them in array-config order, often having to redo the type fix after splitting the function.
3. **Token bloat.** Worst-case mixed failure (5 of 6 categories) is ~5–10K tokens of prompt before the agent's reasoning context starts.
4. **Loss of priority signal.** No way for the prompt to communicate that some fixes block others.

## Proposed fix — small, scoped to the prompt builder

Group findings by category in the rendered prompt with explicit priority preamble. No changes outside `rectifier-builder.ts`. No flow changes.

### Categorization

| Priority | Category bucket | Members | Repair preamble |
|---|---|---|---|
| 1 | Compile/build | `typecheck`, `build` | "Must fix first. Fixes here may resolve later items automatically." |
| 2 | Lint/style | `lint` | "Fix after compile errors. Often automated." |
| 3 | Behavior | `test` | "Fix after compile passes. Indicates implementation does not match assertion." |
| 4 | Semantic | `semantic` | "Story-alignment concerns. Re-read the story before editing." |
| 5 | Architectural | `adversarial` | "Lowest priority. Fix only after all above pass — fixes here may invalidate later if compile errors return." |

The order matches the cost-and-cascade reasoning from #812: cheap before expensive, mechanical before judgment, narrow scope before wide scope.

### Prompt shape (after change)

```
Review failed after your implementation. Fix the following issues (3 attempts available before escalation).

**Order matters: fix Priority 1 first; later priorities may need redo if earlier ones change the code.**

## Priority 1 — Compile/build (must fix first)

### typecheck (exit 2)
```
error TS2688: Cannot find type definition file for 'bun-types'.
```

## Priority 2 — Lint/style

### lint (exit 1)
```
…
```

## Priority 4 — Semantic (story-alignment)

### semantic (exit 1)
- [error] src/foo.ts:42 — Implementation does not satisfy AC#3

## Priority 5 — Architectural (lowest priority — re-verify after fixing above)

### adversarial (exit 1)
- [warn] src/foo.ts:55 — Function is too long, consider splitting
- [info] src/bar.ts:10 — Missing test for new exported symbol

Fix in priority order. After fixing each priority, re-run the failing check(s) at that level to verify.
```

If a priority bucket has no findings, omit the section entirely.

## Acceptance criteria

- [ ] `firstAttemptDelta()` groups findings into the 5 priority buckets above; renders only buckets that have findings; emits the priority preamble at the top
- [ ] `continuation()` mirrors the same grouping for retry attempts
- [ ] If only one priority bucket has findings, the section header degrades gracefully (still useful, not awkward)
- [ ] Snapshot tests in `test/unit/prompts/rectifier-builder.test.ts` cover: single-category, two-categories, all-categories
- [ ] No changes to `src/pipeline/stages/autofix-agent.ts` or any other file outside the builder

## Out of scope (parent: #815)

- Skipping LLM checks when mechanical fail (see #817 — the review-stage gating)
- Separate rectification cycles per phase (see #815 — the two-phase architecture)
- Reviewer-specific CLARIFY routing (see #815)

## Why ship this first

Smallest possible change, no flow restructure, no cost optimization risks. Captures the cognitive-flattening benefit immediately. Survives unchanged when #817 and #815 land — they extend this rendering, not replace it.

## References

- Cost analysis and architectural reasoning: discussion in #812
- Prompt audit evidence: dogfood run `2026-04-29T13-04-25` (audit file `1777467865587-…-rectification-t01.txt`)
- Builder convention SSOT: `.claude/rules/forbidden-patterns.md` § Prompt Builder Convention

--- a/src/prompts/builders/rectifier-builder.ts
+++ b/src/prompts/builders/rectifier-builder.ts
@@ -14,7 +14,7 @@
 
 import type { RectificationConfig } from "../../config";
 import type { UserStory } from "../../prd";
-import type { ReviewCheckResult } from "../../review/types";
+import type { ReviewCheckName, ReviewCheckResult } from "../../review/types";
 import { formatFailureSummary } from "../../verification/parser";
 import type { TestFailure } from "../../verification/types";
 import { priorFailuresSection, universalConstitutionSection, universalContextSection } from "../core";
@@ -40,6 +40,121 @@ export type { FailureRecord, ReviewFinding };
  */
 export type RectifierTrigger = "tdd-test-failure" | "tdd-suite-failure" | "verify-failure" | "review-findings";
 
+type RectificationPriority = "compile-build" | "lint" | "behavior" | "semantic" | "architectural";
+
+interface PriorityBucket {
+  priority: number;
+  heading: string;
+  guidance: string;
+}
+
+const PRIORITY_BUCKETS: Readonly<Record<RectificationPriority, PriorityBucket>> = {
+  "compile-build": {
+    priority: 1,
+    heading: "Compile/build",
+    guidance: "Must fix first. Fixes here may resolve later items automatically.",
+  },
+  lint: {
+    priority: 2,
+    heading: "Lint/style",
+    guidance: "Fix after compile errors. Often automated.",
+  },
+  behavior: {
+    priority: 3,
+    heading: "Behavior",
+    guidance: "Fix after compile passes. Indicates implementation does not match assertion.",
+  },
+  semantic: {
+    priority: 4,
+    heading: "Semantic",
+    guidance: "Story-alignment concerns. Re-read the story before editing.",
+  },
+  architectural: {
+    priority: 5,
+    heading: "Architectural",
+    guidance: "Lowest priority. Fix only after all above pass.",
+  },
+};
+
+const PRIORITY_ORDER: Readonly<RectificationPriority[]> = [
+  "compile-build",
+  "lint",
+  "behavior",
+  "semantic",
+  "architectural",
+];
+
+function priorityForCheck(checkName: ReviewCheckName): RectificationPriority {
+  switch (checkName) {
+    case "typecheck":
+    case "build":
+      return "compile-build";
+    case "lint":
+      return "lint";
+    case "test":
+      return "behavior";
+    case "semantic":
+      return "semantic";
+    case "adversarial":
+      return "architectural";
+    default:
+      return assertNever(checkName);
+  }
+}
+
+function assertNever(value: never): never {
+  throw new Error(`Unhandled review check category: ${String(value)}`);
+}
+
+function renderCheckBlock(check: ReviewCheckResult): string {
+  const parts: string[] = [];
+  parts.push(`### ${check.check} (exit ${check.exitCode})\n`);
+  const truncated = check.output.length > 4000;
+  const output = truncated
+    ? `${check.output.slice(0, 4000)}\n... (truncated — ${check.output.length} chars total)`
+    : check.output;
+  parts.push(`\`\`\`\n${output}\n\`\`\`\n`);
+
+  if (check.findings?.length) {
+    parts.push("Structured findings:\n");
+    for (const f of check.findings) {
+      parts.push(`- [${f.severity}] ${f.file}:${f.line} — ${f.message}\n`);
+    }
+  }
+
+  return parts.join("\n");
+}
+
+function renderPrioritizedFailures(failedChecks: Readonly<ReviewCheckResult[]>): string {
+  const grouped: Readonly<Record<RectificationPriority, Readonly<ReviewCheckResult[]>>> = {
+    "compile-build": failedChecks.filter((check) => priorityForCheck(check.check) === "compile-build"),
+    lint: failedChecks.filter((check) => priorityForCheck(check.check) === "lint"),
+    behavior: failedChecks.filter((check) => priorityForCheck(check.check) === "behavior"),
+    semantic: failedChecks.filter((check) => priorityForCheck(check.check) === "semantic"),
+    architectural: failedChecks.filter((check) => priorityForCheck(check.check) === "architectural"),
+  };
+
+  const sections: string[] = [
+    "**Order matters: fix Priority 1 first; later priorities may need redo if earlier ones change the code.**\n",
+  ];
+
+  for (const priority of PRIORITY_ORDER) {
+    const checks = grouped[priority];
+    if (checks.length === 0) {
+      continue;
+    }
+
+    const bucket = PRIORITY_BUCKETS[priority];
+    sections.push(`## Priority ${bucket.priority} — ${bucket.heading}\n`);
+    sections.push(`${bucket.guidance}\n`);
+    for (const check of checks) {
+      sections.push(renderCheckBlock(check));
+    }
+  }
+
+  return sections.join("\n");
+}
+
 // biome-ignore lint/complexity/noStaticOnlyClass: Static-method namespace for prompt builders (ADR-018)
 export class RectifierPromptBuilder {
   /**
@@ -58,23 +173,10 @@ export class RectifierPromptBuilder {
       `Review failed after your implementation. Fix the following issues (${attemptWord} available before escalation):\n`,
     );
 
-    for (const check of failedChecks) {
-      parts.push(`### ${check.check} (exit ${check.exitCode})\n`);
-      const truncated = check.output.length > 4000;
-      const output = truncated
-        ? `${check.output.slice(0, 4000)}\n... (truncated — ${check.output.length} chars total)`
-        : check.output;
-      parts.push(`\`\`\`\n${output}\n\`\`\`\n`);
-      if (check.findings?.length) {
-        parts.push("Structured findings:\n");
-        for (const f of check.findings) {
-          parts.push(`- [${f.severity}] ${f.file}:${f.line} — ${f.message}\n`);
-        }
-      }
-    }
+    parts.push(renderPrioritizedFailures(failedChecks));
 
     parts.push(
-      "\nFix ALL issues listed. After fixing, re-run the failing check(s) to verify they pass before committing. Do NOT change test files or test behavior. Commit your changes when all checks pass.",
+      "\nFix in priority order. After fixing each priority, re-run the failing check(s) at that level to verify they pass before moving on. Do NOT change test files or test behavior. Commit your changes when all checks pass.",
     );
     parts.push(CONTRADICTION_ESCAPE_HATCH);
 
@@ -98,20 +200,7 @@ export class RectifierPromptBuilder {
 
     parts.push("Your previous fix attempt did not resolve all issues. Here are the remaining failures:\n");
 
-    for (const check of failedChecks) {
-      parts.push(`### ${check.check} (exit ${check.exitCode})\n`);
-      const truncated = check.output.length > 4000;
-      const output = truncated
-        ? `${check.output.slice(0, 4000)}\n... (truncated — ${check.output.length} chars total)`
-        : check.output;
-      parts.push(`\`\`\`\n${output}\n\`\`\`\n`);
-      if (check.findings?.length) {
-        parts.push("Structured findings:\n");
-        for (const f of check.findings) {
-          parts.push(`- [${f.severity}] ${f.file}:${f.line} — ${f.message}\n`);
-        }
-      }
-    }
+    parts.push(renderPrioritizedFailures(failedChecks));
 
     if (attempt >= rethinkAtAttempt) {
       parts.push(

--- a/test/unit/prompts/builders/rectifier-builder.test.ts
+++ b/test/unit/prompts/builders/rectifier-builder.test.ts
@@ -136,14 +136,14 @@ describe("RectifierPromptBuilder.firstAttemptDelta", () => {
     expect(prompt).toContain("UNRESOLVED:");
   });
 
-  test("instructs agent to fix ALL issues, verify, and commit", () => {
+  test("instructs agent to fix in priority order, verify, and commit", () => {
     const prompt = RectifierPromptBuilder.firstAttemptDelta(
       [makeCheck("lint", "error")],
       2,
     );
 
-    expect(prompt).toContain("Fix ALL issues listed");
-    expect(prompt).toContain("re-run the failing check(s) to verify they pass before committing");
+    expect(prompt).toContain("Fix in priority order");
+    expect(prompt).toContain("re-run the failing check(s) at that level to verify they pass before moving on");
     expect(prompt).toContain("Commit your changes when all checks pass");
   });
 

--- a/test/unit/prompts/rectifier-builder.test.ts
+++ b/test/unit/prompts/rectifier-builder.test.ts
@@ -12,6 +12,7 @@ import { describe, expect, test } from "bun:test";
 import { makeStory } from "../../helpers";
 import { RectifierPromptBuilder } from "../../../src/prompts";
 import type { FailureRecord } from "../../../src/prompts";
+import type { ReviewCheckResult } from "../../../src/review/types";
 
 // ─── Fixtures ─────────────────────────────────────────────────────────────────
 
@@ -224,5 +225,148 @@ describe("RectifierPromptBuilder.noOpReprompt", () => {
     const atLimit = RectifierPromptBuilder.noOpReprompt([FAILED_CHECK], 1, 1);
     expect(beforeLimit).not.toContain("WARNING");
     expect(atLimit).toContain("WARNING");
+  });
+});
+
+// ─── firstAttemptDelta / continuation priority-bucket rendering ───────────────
+
+const makeReviewCheck = (
+  check: ReviewCheckResult["check"],
+  overrides: Partial<ReviewCheckResult> = {},
+): ReviewCheckResult => {
+  return {
+    check,
+    success: false,
+    command: `${check} command`,
+    exitCode: 1,
+    output: `${check} output`,
+    durationMs: 10,
+    ...overrides,
+  };
+};
+
+describe("RectifierPromptBuilder.firstAttemptDelta", () => {
+  test("single-category: renders only the matching priority bucket", () => {
+    const result = RectifierPromptBuilder.firstAttemptDelta([makeReviewCheck("lint")], 3);
+
+    expect(result).toContain("Order matters: fix Priority 1 first");
+    expect(result).toContain("## Priority 2 — Lint/style");
+    expect(result).not.toContain("## Priority 1 — Compile/build");
+    expect(result).not.toContain("## Priority 3 — Behavior");
+    expect(result).not.toContain("## Priority 4 — Semantic");
+    expect(result).not.toContain("## Priority 5 — Architectural");
+    expect(result).toMatchSnapshot();
+  });
+
+  test("two-categories: renders in fixed priority order, not input order", () => {
+    const result = RectifierPromptBuilder.firstAttemptDelta(
+      [makeReviewCheck("semantic"), makeReviewCheck("typecheck")],
+      3,
+    );
+
+    expect(result).toContain("## Priority 1 — Compile/build");
+    expect(result).toContain("## Priority 4 — Semantic");
+    expect(result.indexOf("## Priority 1 — Compile/build")).toBeLessThan(result.indexOf("## Priority 4 — Semantic"));
+    expect(result).toMatchSnapshot();
+  });
+
+  test("all-categories: renders all five buckets with expected grouping", () => {
+    const result = RectifierPromptBuilder.firstAttemptDelta(
+      [
+        makeReviewCheck("adversarial"),
+        makeReviewCheck("build", { exitCode: 2 }),
+        makeReviewCheck("semantic", {
+          findings: [
+            {
+              ruleId: "semantic-ac3",
+              severity: "error",
+              file: "src/foo.ts",
+              line: 42,
+              message: "Implementation does not satisfy AC#3",
+            },
+          ],
+        }),
+        makeReviewCheck("test"),
+        makeReviewCheck("lint"),
+        makeReviewCheck("typecheck"),
+      ],
+      3,
+    );
+
+    expect(result).toContain("## Priority 1 — Compile/build");
+    expect(result).toContain("### typecheck (exit 1)");
+    expect(result).toContain("### build (exit 2)");
+    expect(result).toContain("## Priority 2 — Lint/style");
+    expect(result).toContain("## Priority 3 — Behavior");
+    expect(result).toContain("## Priority 4 — Semantic");
+    expect(result).toContain("## Priority 5 — Architectural");
+    expect(result).toContain("Structured findings:");
+    expect(result).toMatchSnapshot();
+  });
+});
+
+describe("RectifierPromptBuilder.continuation", () => {
+  test("single-category: renders only the matching priority bucket", () => {
+    const result = RectifierPromptBuilder.continuation([makeReviewCheck("lint")], 1, 2, 3);
+
+    expect(result).toContain("Order matters: fix Priority 1 first");
+    expect(result).toContain("## Priority 2 — Lint/style");
+    expect(result).not.toContain("## Priority 1 — Compile/build");
+    expect(result).not.toContain("## Priority 3 — Behavior");
+    expect(result).not.toContain("## Priority 4 — Semantic");
+    expect(result).not.toContain("## Priority 5 — Architectural");
+    expect(result).toMatchSnapshot();
+  });
+
+  test("two-categories: renders in fixed priority order, not input order", () => {
+    const result = RectifierPromptBuilder.continuation(
+      [makeReviewCheck("semantic"), makeReviewCheck("typecheck")],
+      1,
+      2,
+      3,
+    );
+
+    expect(result).toContain("## Priority 1 — Compile/build");
+    expect(result).toContain("## Priority 4 — Semantic");
+    expect(result.indexOf("## Priority 1 — Compile/build")).toBeLessThan(result.indexOf("## Priority 4 — Semantic"));
+    expect(result).toMatchSnapshot();
+  });
+
+  test("all-categories: renders all five buckets with expected grouping", () => {
+    const result = RectifierPromptBuilder.continuation(
+      [
+        makeReviewCheck("adversarial"),
+        makeReviewCheck("build", { exitCode: 2 }),
+        makeReviewCheck("semantic", {
+          findings: [
+            {
+              ruleId: "semantic-ac3",
+              severity: "error",
+              file: "src/foo.ts",
+              line: 42,
+              message: "Implementation does not satisfy AC#3",
+            },
+          ],
+        }),
+        makeReviewCheck("test"),
+        makeReviewCheck("lint"),
+        makeReviewCheck("typecheck"),
+      ],
+      3,
+      2,
+      3,
+    );
+
+    expect(result).toContain("## Priority 1 — Compile/build");
+    expect(result).toContain("### typecheck (exit 1)");
+    expect(result).toContain("### build (exit 2)");
+    expect(result).toContain("## Priority 2 — Lint/style");
+    expect(result).toContain("## Priority 3 — Behavior");
+    expect(result).toContain("## Priority 4 — Semantic");
+    expect(result).toContain("## Priority 5 — Architectural");
+    expect(result).toContain("Structured findings:");
+    expect(result).toContain("Rethink your approach");
+    expect(result).toContain("URGENT: This is your final attempt");
+    expect(result).toMatchSnapshot();
   });
 });

